pkg/controller/agent/controller.go

```diff
@@ -157,19 +160,54 @@ func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Ag
 		return results
 	}
 
-	// Run basic validations as a fallback in case webhook is disabled.
-	if err := r.validate(ctx, agent); err != nil {
-		return results.WithError(err)
-	}
-
-	driverResults := internalReconcile(Params{
+	params := Params{
 		Context:        ctx,
 		Client:         r.Client,
 		EventRecorder:  r.recorder,
 		Watches:        r.dynamicWatches,
 		Agent:          agent,
 		OperatorParams: r.Parameters,
-	})
+	}
+
+	status := newStatus(agent)
+
+	// Run basic validations as a fallback in case webhook is disabled.
+	if err := r.validate(ctx, agent); err != nil {
+		if statusErr := updateStatus(params, status); statusErr != nil && apierrors.IsConflict(statusErr) {
+			params.Logger().V(1).Info(
+				"Conflict while updating status",
+				"namespace", params.Agent.Namespace,
+				"agent_name", params.Agent.Name)
+			return results.WithResult(reconcile.Result{Requeue: true})
+		} else if err != nil {
+			params.Logger().Error(
+				err,
+				"Error while updating status",
+				"namespace", params.Agent.Namespace,
+				"agent_name", params.Agent.Name,
+			)
+			return results.WithError(errors.Wrapf(err, "while updating status: %s", statusErr))
```

Comment:
```golang
func (r *ReconcileAgent) doReconcile(ctx context.Context, agent agentv1alpha1.Agent) (results *reconciler.Results) {
	defer tracing.Span(&ctx)()
        status := newStatus(agent)
	results = reconciler.NewResult(ctx)
	if !association.AreConfiguredIfSet(agent.GetAssociations(), r.recorder) {
		return results
	}

	params := Params{
		Context:        ctx,
		Client:         r.Client,
		EventRecorder:  r.recorder,
		Watches:        r.dynamicWatches,
		Agent:          agent,
		OperatorParams: r.Parameters,
	}	

	defer func() {
		if err := updateStatus(params, status); err != nil && apierrors.IsConflict(err) {
			params.Logger().V(1).Info(
				"Conflict while updating status",
				"namespace", params.Agent.Namespace,
				"agent_name", params.Agent.Name)
			results = results.WithResult(reconcile.Result{Requeue: true})
		} else if err != nil {
			params.Logger().Error(
				err,
				"Error while updating status",
				"namespace", params.Agent.Namespace,
				"agent_name", params.Agent.Name,
			)
			results = results.WithError(err)
		}
	}()

	// Run basic validations as a fallback in case webhook is disabled.
	if err := r.validate(ctx, agent); err != nil {
		return results.WithError(err)
	}

	return results.WithResults(internalReconcile(params, &status))
}
```

Reply:
I went with deferring, as you suggested here.  Makes sense.